addons/web/static/src/js/core/owl_popover.js

```diff
@@ -0,0 +1,234 @@
+            that fits the viewport) if it fits the viewport, otherwise
+            fallback to 'bottom' if no position fits the viewport */
+            if (position) {
+                this.popoverRef.el.classList.add(`o_popover--${position.name}`);
+            } else {
+                this.popoverRef.el.style.top = `${positioningData.bottom.top}px`;
+                this.popoverRef.el.style.left = `${positioningData.bottom.left}px`;
+                this.popoverRef.el.classList.add(`o_popover--bottom`);
+            }
+        }
+
+        //--------------------------------------------------------------------------
+        // Private
+        //--------------------------------------------------------------------------
+
+        /**
+         * Send an event signaling that the popover must be closed.
+         * @private
+         */
+        _close() {
+            this.trigger('close-popover');
+        }
+
+        /**
+         * Check if the element is in the viewport.
+         * @param {HTMLElement}
+         * @returns {Boolean} True if the popover fits in the viewport, false otherwise.
+         */
+        _isInViewport(element) {
+            const rect = element.getBoundingClientRect();
+            const html = document.documentElement;
+            return (
+                rect.top >= 0 &&
+                rect.left >= 0 &&
+                rect.bottom <= (window.innerHeight || html.clientHeight) &&
+                rect.right <= (window.innerWidth || html.clientWidth)
+            );
+        }
+
+        //--------------------------------------------------------------------------
+        // Handlers
+        //--------------------------------------------------------------------------
+        /**
+         * Handle clicks outside of the popover to trigger its closing.
+         * Note: if the click event is stopped before bubbling up to the document,
+         * this handler will *not* be fired. Make sure to add custom click handlers
+         * if necessary.
+         * @private
+         * @param {DOMEvent} e
+         */
+        _onDocumentClick(e) {
+            if (!this.popoverRef.el.contains(e.target)) {
+                this._close();
+            }
+        }
+
+        /**
+         * Handle changes in the window size to reposition the popover next to
+         * its target.
+         * @private
+         * @param {DOMEvent} e
+         */
+        _onTargetMoved(e) {
+            this.display();
+        }
+
+        //--------------------------------------------------------------------------
+        // Static
+        //--------------------------------------------------------------------------
+
+        /**
+         * Compute the expected positioning coordinates for each possible
+         * positioning based on the target and popover sizes.
+         * @returns {Object}
+         */
+        static _computePositioningData(popoverElement, targetElement) {
+            const {
```

Comment:
I'm not comfi with that style, but that's only personal opinion
I'd go for `const {x, y} = obj` on one line, even if the line is a bit long

Reply:
that's copy/paste code (no shame, check the gist in the PR description :p), i can reformat it for more readability without problem